### ISSUE TITLE
Twitter scrapes fail with `TypeError: __init__() missing 1 required positional argument: 'thumbnailUrl'` (or `'title'`)

### ISSUE THREAD

**Issue by martinratinaud** [JustAnotherArchivist/snscrape#407]
Last commit seems to break the execution of `snscrape` https://github.com/JustAnotherArchivist/snscrape/commit/25ee014e2970737e500e2964ab09584b11b19fb0

I get ` missing 1 required positional argument: 'thumbnailUrl'` when using

`/home/ambnum/.local/bin/snscrape --with-entity --max-results 1000 --jsonl twitter-search "+@nath_yamb " > "/tmp/information-manipulation-analyzer/@nath_yamb/original.json"`

**Comment by JustAnotherArchivist** [comment]
Interesting, thank you. I searched for different types of cards but didn't come across this one. The offending tweet is 1493756926869397504, and it also crashes on the non-search scrapers.

**Comment by hinceme** [comment]
Hi there, I'm having a similar issue when using a TwitterSearchSraper:

`Traceback (most recent call last):
  File "./scrape.py", line 35, in <module>
    for k, tweet in enumerate(sntwitter.TwitterSearchScraper(terms[j] + ' since:2017-01-01 until:2020-12-31 near:' + "\"" + locations_stripped[i] + "\"" + 'within:30mi').get_items()):
  File "/home/ec2-user/.local/lib/python3.8/site-packages/snscrape/modules/twitter.py", line 797, in get_items
    yield from self._v2_timeline_instructions_to_tweets(obj)
  File "/home/ec2-user/.local/lib/python3.8/site-packages/snscrape/modules/twitter.py", line 483, in _v2_timeline_instructions_to_tweets
    yield from self._v2_instruction_tweet_entry_to_tweet(entry['entryId'], entry['content'], obj)
  File "/home/ec2-user/.local/lib/python3.8/site-packages/snscrape/modules/twitter.py", line 506, in _v2_instruction_tweet_entry_to_tweet
    yield self._tweet_to_tweet(tweet, obj)
  File "/home/ec2-user/.local/lib/python3.8/site-packages/snscrape/modules/twitter.py", line 630, in _tweet_to_tweet
    kwargs['card'] = self._make_card(tweet['card'], _TwitterAPIType.V2)
  File "/home/ec2-user/.local/lib/python3.8/site-packages/snscrape/modules/twitter.py", line 620, in _make_card
    return Card(**cardKwargs)
TypeError: __init__() missing 1 required positional argument: 'description'`

If there's any other information you may need from me, please let me know! Thank you in advance!

**Comment by JustAnotherArchivist** [comment]
@hinceme Thanks! What's the search query that triggered this?

Also, if anyone has one with an error for `title` or `url`, I'd be interested in that as well.

**Comment by hinceme** [comment]
I'm running a list of search terms, and it seems to get stuck on "New World Order" as the term. Here's the entire query: 

`sntwitter.TwitterSearchScraper(terms[j] + ' since:2017-01-01 until:2020-12-31 near:' + "\"" + locations_stripped[i] + "\"" + 'within:30mi'` where `terms[j]` is "New World Order" and `locations_stripped[i]` is Minneapolis, MN

**Comment by JustAnotherArchivist** [comment]
Hmm, interesting, I'm getting the error for `thumbnailUrl` with that, not `description`. Could you try and reproduce it with the CLI and, if successful, tell me the `cardKwargs['url']` near the end of the dump file (filename will appear in a `CRITICAL` log message just before the crash)?

**Comment by hinceme** [comment]
I will try my best and get back to you. Thank you for taking the time to look into this!

**Comment by JustAnotherArchivist** [comment]
Actually, it looks like Twitter's search is somewhat broken at the moment. I can't even reliably reproduce the crash with the same query, only every few attempts... I'll let it run for a while and see if I can crash it with `description`.

**Comment by JustAnotherArchivist** [comment]
All good, I found another reproducer for the `description` crash: 1493995454043992069

**Comment by JustAnotherArchivist** [comment]
Fixed for `thumbnailUrl` and `description`.

As mentioned, if anyone sees this with `title` or `url` instead, please let me know.

**Comment by hinceme** [comment]
You are amazing, thank you so much!

**Comment by akunGB123** [comment]
> Fixed for `thumbnailUrl` and `description`.
> 
> As mentioned, if anyone sees this with `title` or `url` instead, please let me know.

i have it sir. it says :
`TypeError: __init__() missing 1 required positional argument: 'title'`

**Comment by hinceme** [comment]
> Fixed for `thumbnailUrl` and `description`.
> 
> As mentioned, if anyone sees this with `title` or `url` instead, please let me know.

I am actually now getting this error as well. My query is `sntwitter.TwitterSearchScraper(terms[j] + ' since:2017-01-01 until:2020-12-31 near:' + "\"" + locations_stripped[i] + "\"" + 'within:30mi').get_items()` where `terms[j]` is "joo" and `locations_stripped[i]` is Minneapolis, MN

**Comment by JustAnotherArchivist** [comment]
@akunGB123 Thanks! The culprit is 1334631272715575300. I had no idea that polls were also cards internally...

@hinceme I can't reproduce that, but I'm getting unreliable results anyway. It's probably the same thing as last night. (By the way, there should be a space before `within`.)

**Comment by hinceme** [comment]
> @akunGB123 Thanks! The culprit is 1334631272715575300. I had no idea that polls were also cards internally...
> 
> @hinceme I can't reproduce that, but I'm getting unreliable results anyway. It's probably the same thing as last night. (By the way, there should be a space before `within`.)

Ah thank you, will fix that! And I'm not sure if it matters but I'm running my code on an AWS Linux instance, I don't think that would make a difference in reproducing errors but I figured I'd mention it just in case.

**Comment by TheTechRobo** [comment]
Isn't AWS blocked by Twitter?

**Comment by JustAnotherArchivist** [comment]
Interesting that it's working at all for you then since Twitter blocks AWS IPs (#79). It's probably an edge case where it depends on which search server you hit. I'm getting zero, two, or three results from Europe and no crashes.

**Comment by hinceme** [comment]
That's interesting, I didn't know AWS was blocked by Twitter. I'm currently a college student so I think I should be able to use the student network to distribute this and do the work that way. I feel stupid that I didn't know AWS was blocked

**Comment by JustAnotherArchivist** [comment]
I mean, if you're getting results, it's clearly not entirely blocked. Or perhaps the block has been lifted since. Anyway, that's off-topic for this issue. :-)

**Comment by Ori-Pixel** [comment]
Getting the title issue with:

```
for i, tweet in enumerate(sntwitter.TwitterSearchScraper('from:jack').get_items()):
    if i > 100:
        break
    tweets_list1.append([tweet.date, tweet.id, tweet.content, tweet.user.username])
```
`TypeError: Card.__init__() missing 1 required positional argument: 'title'`

**Comment by JustAnotherArchivist** [comment]
Thanks! That's yet another card type on tweet 1494017100888870916.

I'll have to come up with a more robust way of handling this.

**Comment by Ori-Pixel** [comment]
I ended up fixing it in the same manner with the optional typing on title and it works:

```
class Card:
	url: str
	title: typing.Optional[str] = None
	description: typing.Optional[str] = None
	thumbnailUrl: typing.Optional[str] = None
```

**Comment by JustAnotherArchivist** [comment]
Yes, but I'd like to actually extract the available information, e.g. poll and audio details in those two examples, if easily possible. I'll likely end up checking the card type and emitting a warning if something is unsupported.

**Comment by Ori-Pixel** [comment]
All good. I'll wait for an update for main.

**Comment by JustAnotherArchivist** [comment]
Resolving this properly will take a bit. It turns out that there are over two dozen different card types, even more if you count several very similar ones individually, and some with subtypes. I have found examples for most of them.

**Comment by JustAnotherArchivist** [comment]
A little progress update on this: I've implemented all card types for which I could find examples except `unified_card` (which is more complicated). There are a few that I simply couldn't find in use because the search doesn't seem to return them; from the names, some of them are only relevant to direct messages, but others should probably be present in public tweets. Those will emit a warning for now, which can be resolved at a later time.

**Comment by bndr** [comment]
I'm also currently having this issue with 'title'. Looking forward to the update.

**Comment by mlcaepiee** [comment]
Hi, I have met the problem with 'TypeError: __init__() missing 1 required positional argument: 'title''.

My query is：
`    for i, tweet in enumerate(sntwitter.TwitterSearchScraper('Beijing2022 lang:en filter:hashtags since:2022-02-04 '
                                                             'until:2022-02-05').get_items()):
        if i > 200:
            break
        print(i)`
the loop will crush when i = 196

Looking forward to the update and thank you.

**Comment by k-marcelino** [comment]
> Fixed for `thumbnailUrl` and `description`.
> 
> As mentioned, if anyone sees this with `title` or `url` instead, please let me know.

Hello! Untill yesterday I was extracting data perfectly from a twitter user, today this 'title' error started happening.
Here's the code:

![image](https://user-images.githubusercontent.com/85710050/156235400-43adca9d-040e-4a20-a618-7cea3c756ff3.png)

Here's the error:

![image](https://user-images.githubusercontent.com/85710050/156235568-41b814e7-98f1-470c-9e63-be6d084b9311.png)

I tried to reinstall scscrape and development version again, but it continues to happen, Thanks!!!

**Comment by raoumairwaheed** [comment]
TypeError: __init__() missing 1 required positional argument: 'title'

Hello Sir, I am also getting this error. 
my code is: 

for i, tweet in enumerate(sntwitter.TwitterSearchScraper('from:Nike').get_items()):  # declare a username
    if i > 100:  # number of tweets you want to scrape
        break
    if tweet.media:
        print(tweet.id, tweet.url,  tweet.media)

**Comment by AyushiG1** [comment]
I am also getting this error
![image](https://user-images.githubusercontent.com/40836951/157233937-4d642a89-56cd-4837-a427-ef9dc45bebb8.png)

**Comment by hwhjones** [comment]
![image](https://user-images.githubusercontent.com/77008848/158125544-42d96c80-3191-4add-be38-a0c60f08ca0a.png)

![image](https://user-images.githubusercontent.com/77008848/158125880-272a9204-73bd-4f16-8b3e-123d8815d858.png)

Hi there - I know this is an open ticket, but it was a bit weird, I was running TwitterSearchScraper for 10000s of tweets prior to the last couple of days - I now almost always get this error for any search term.

**Comment by TheTechRobo** [comment]
It's really luck whether you get the error or not.

Some search terms have old tweets that won't have these cards, some search terms have new tweets but don't really have these cards either, etc.

**Comment by sharonwu827** [comment]
Hello i also ran into the issus saying "missing title" any updates for solve it?

**Comment by JustAnotherArchivist** [comment]
No update since my last comment here. I've been preoccupied with other things mostly for the past couple weeks. I hope to get this finalised soon though.

**Comment by NoraAlt** [comment]
I have the same error: `TypeError: __init__() missing 1 required positional argument: 'title'`

Any updates to solve this?

**Comment by holma91** [comment]
getting `TypeError:` Card.__init__() missing 1 required positional argument: 'title'` when retrieving this tweet https://twitter.com/jack/status/1494017100888870916 . I assume that it's because a post of a twitter space.

**Comment by holma91** [comment]
> getting `TypeError: Card.__init__() missing 1 required positional argument: 'title'` when retrieving this tweet https://twitter.com/jack/status/1494017100888870916 . I assume that it's because a post of a twitter space.

So I have a hacky solution to this. Changes at the the in function _make_card in snscrape/modules/twitter.py:
```
### ADDED BY ME! START
if 'title' not in cardKwargs:
    cardKwargs['title'] = 'THIS IS A TWEET LINKING A TWITTER SPACE'
### ADDED BY ME! END
return Card(**cardKwargs)
```

if you later get `KeyError: 'tcooutlinks'`, you can use another hacky solution. In the end of function _make_tweet:
```
    try:
        i = kwargs['tcooutlinks'].index(card.url)
    except ValueError:
         _logger.warning('Could not find card URL in tcooutlinks')
### ADDED BY ME! START
    except KeyError:
         _logger.warning('Could not find key tcooutlinks')
### ADDED BY ME! END
    else:
        card.url = kwargs['outlinks'][i]
return Tweet(**kwargs)
```

Keep in mind that these are only temporary hacks to go around the current problem with twitter spaces.

**Comment by DtneSEffct** [comment]
I too am getting the 'title' related error on another machine, but it is running without any error in my personal machine.

**Comment by JustAnotherArchivist** [comment]
@DtneSEffct You probably have different versions installed. This issue only affects the dev version.

**Comment by JustAnotherArchivist** [comment]
For the same scrapes? If not, maybe you're just getting lucky on one of the machines. This bug is only triggered when you hit certain tweets (namely ones that have a non-summary card), of which there are relatively few.

**Comment by MurasatoNaoya** [comment]
Thank you for your hard work. I updated snscrape to account for your recent commit, and while I am not getting any more 'title' errors, I am instead sometimes getting 'missing 1 required positional argument: 'thumbnailUrl'' errors. Could I be doing something wrong on my part, or is there maybe something else going on? 

Thank you. 

`
tweet_list = []

keyword = str(input("Please enter a keyword or hashtag to focus your search: "))
NoOfTweets = int(input("Please enter the number of Tweets you would like to analyse: "))


for i,tweet in enumerate(sntwitter.TwitterSearchScraper(f'{keyword} near:"Greater Manchester" within:50km + since:2021-02-16 until:2022-3-8').get_items()):
    if i>=NoOfTweets:
        break
    tweet_list.append([tweet.content])
    

df = pd.DataFrame(tweet_list, columns=['Text'])`

**Comment by JustAnotherArchivist** [comment]
@MurasatoNaoya No, you're not doing anything wrong. I've already seen this as well and fixed it locally. Twitter changed things again in the past few weeks... More fixes coming today.

**Comment by TheTechRobo** [comment]
By the way, @MurasatoNaoya I don't think you need to specifically call `str` on an `input` call.

**Comment by MurasatoNaoya** [comment]
I see, thanks for the fast update. If there is anywhere you accept donations or contributions please say, the work you do here is pretty amazing. Cheers.

**Comment by MurasatoNaoya** [comment]
> By the way, @MurasatoNaoya I don't think you need to specifically call `str` on an `input` call.

Thank for the tip, I'll update that now. I'm new to programming, so I really need all the help I can get!

**Comment by JustAnotherArchivist** [comment]
All known issues with cards should be fixed by the current master (3870282a). There's always a chance more might surface later. Please file separate issues for those if you come across any.
